spring-cloud-aws-messaging/src/main/java/org/springframework/cloud/aws/messaging/listener/QueueMessageHandler.java

```diff
@@ -88,6 +92,44 @@ public QueueMessageHandler() {
 		return new String[] { valueToWrap.toString() };
 	}
 
+	@Override
+	public void handleMessage(Message<?> message) throws MessagingException {
+		String destination = getDestination(message);
+		if (destination == null) {
+			return;
+		}
+		String lookupDestination = getLookupDestination(destination);
+		if (lookupDestination == null) {
+			return;
+		}
+
+		MessageHeaderAccessor headerAccessor = MessageHeaderAccessor
+				.getMutableAccessor(message);
+		headerAccessor.setHeader(
+				DestinationPatternsMessageCondition.LOOKUP_DESTINATION_HEADER,
+				lookupDestination);
+		headerAccessor.setLeaveMutable(true);
+
+		if (message instanceof SqsMessage) {
+			message = new SqsMessage((String) message.getPayload(),
```

Comment:
This whole method had to be copied from parent class except this message creation. Parent class `handleMessage` ignores the type of original message and creates just a new `GenericMessage` - meaning we lose the object created in `QueueMessageUtils`.

Reply:
Have You maybe considered something like this:
1. Do not modify QueueMessageHandler by overriding handleMessage. 
2. In QueueMessageUtils change method: 
`	public static Message<String> createMessage(
			com.amazonaws.services.sqs.model.Message message,
			Map<String, Object> additionalHeaders) {
		HashMap<String, Object> messageHeaders = new HashMap<>();
		messageHeaders.put(MESSAGE_ID_MESSAGE_ATTRIBUTE_NAME, message.getMessageId());
		messageHeaders.put(RECEIPT_HANDLE_MESSAGE_ATTRIBUTE_NAME,
				message.getReceiptHandle());

		messageHeaders.putAll(additionalHeaders);
		messageHeaders.putAll(getAttributesAsMessageHeaders(message));
		messageHeaders.putAll(getMessageAttributesAsMessageHeaders(message));

		return new GenericMessage<>(message.getBody(),
				new SqsMessageHeaders(messageHeaders));
	}`

So that it does not create GenericMessage with String payload, but rather with original message as a payload ? 
Resolving to original message or String can be then done by argument resolver.

In above approach You are not say 'impairing' intentions of AbstractMethodMessageHandler class which comes from  messaging, and You are changing code belonging only to spring-cloud-aws. 

I must say that this is just unconfirmed idea, I didn't check if changing QueueMessageUtils.createMessage to return GenericMessage<...aws.Message> is not breaking anything.....

Reply:
But GenericMessage encapsulating entire original message... still workaround. Maybe not that good as I was thinking.

Reply:
Thanks for an idea! I'll experiment with it and keep you updated